File: Data Structures & Algorithms/buy-and-sell-crypto/submission-0.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        // [10, 1, 9, 2, 3]
        // 10 -> 0
        // 1, -> 0
        // 1, 9 -> 8
        // 1, 2 -> 8
        // 1, 3
        int max_profit = 0;
        int lowest = prices[0];
        for (int i = 1; i < prices.size(); i++) {
            if (prices[i] < lowest) {
                lowest = prices[i];
                continue;
            }
            max_profit = max(max_profit, prices[i] - lowest);
        }
        return max_profit;
    }
};

File: Data Structures & Algorithms/stone-game/submission-0.py
class Solution:
    def stoneGame(self, piles: List[int]) -> bool:
        stones = {}  # stones[(l, r)]: total stones in (left, right) range

        def pick(l, r):
            if (l > r):
                return 0
            if (l, r) in stones:
                return stones[(l, r)]

            if (r - l + 1) % 2 == 1:
                piles[l] = piles[r] = 0

            stones[(l, r)] = max(
                pick(l + 1, r) + piles[l], 
                pick(l, r - 1) + piles[r]
            )

            return stones[(l, r)]

        alice = pick(0, len(piles) - 1)
        return alice > (sum(piles) - alice)
